### PR TITLE
[SPARK-13376] [SPARK-13476] [SQL] improve column pruning

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

This PR mostly rewrite the ColumnPruning rule to support most of the SQL logical plans (except those for Dataset).

This PR also fix a bug in Generate, it should always output UnsafeRow, added an regression test for that.
## How was this patch tested?

This is test by unit tests, also manually test with TPCDS Q78, which could prune all unused columns successfully, improved the performance by 78% (from 22s to 12s).

### CODE REVIEW

**Comment by davies** [comment]
@rxin Please review the final commit 52e98b8

**Comment by rxin** [comment]
LGTM pending tests pass.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala [hunk 1]
```diff
@@ -73,19 +76,26 @@ case class Generate(
           if (outer && outputRows.isEmpty) {
             joinedRow.withRight(generatorNullRow) :: Nil
           } else {
-            outputRows.map(or => joinedRow.withRight(or))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala]: This line is the cause: we forget to do unsafe project here.

**Comment by cloud-fan** [comment]
LGTM

**Comment by davies** [comment]
Merging into master.

**Pull request closed without merging**